kratos/python_interface/python_registry.py

```diff
@@ -0,0 +1,370 @@
+
+        # Dictionary to act as Python registry
+        self.__python_registry = {}
+
+    def __getitem__(self, Name):
+        if self.HasItem(Name, RegistryContext.PYTHON):
+            # Get the class from the Python registry
+            return self.__InternalGetPythonItem(Name)
+        elif self.HasItem(Name, RegistryContext.CPP):
+            # Get the class from the c++ registry
+            return self.__InternalGetCppItem(Name)
+        else:
+            err_msg = f"Asking to retrieve '{Name}' non-registered item."
+            raise Exception(err_msg)
+
+    def __iter__(self):
+       ''' Returns the Iterator object '''
+       return PythonRegistryIterator(self)
+
+    def keys(self, Name):
+        # Check if Name registry level is registered and if it is a value
+        if self.HasItem(Name):
+            if not self.HasItems(Name):
+                err_mgs = f"Asking for the keys of '{Name}'. '{Name}' item has no subitems."
+                raise Exception(err_mgs)
+        else:
+            err_msg = f"Asking for the keys of '{Name}' non-registered item."
+            raise Exception(err_msg)
+
+        # Get the c++ registry item keys
+        cpp_keys = None
+        if self.HasItems(Name, RegistryContext.CPP):
+            cpp_keys = self.__InternalGetCppItem(Name).keys()
+        # Get the Python registry item keys
+        # Note that this is a view object of the Python dictionary keys
+        py_keys = None
+        if self.HasItems(Name, RegistryContext.PYTHON):
+            py_keys = self.__InternalGetPythonItem(Name).keys()
+
+        # Return the c++ and Python registry keys in a tuple
+        # Note that we return None if the item was not present in one of the registries
+        return py_keys, cpp_keys
+
+    def values(self, Name):
+        # Check if Name registry level is registered and if it is a value
+        if self.HasItem(Name):
+            if not self.HasItems(Name):
+                err_mgs = f"Asking for the keys of '{Name}'. '{Name}' item has no subitems."
+                raise Exception(err_mgs)
+        else:
+            err_msg = f"Asking for the keys of '{Name}' non-registered item."
+            raise Exception(err_msg)
+
+        # Get the c++ registry item values
+        cpp_values = None
+        if self.HasItems(Name, RegistryContext.CPP):
+            cpp_values = self.__InternalGetCppItem(Name).values()
+        # Get the Python registry item values
+        # Note that this is a view object of the Python dictionary values
+        py_values = None
+        if self.HasItems(Name, RegistryContext.PYTHON):
+            py_values = self.__InternalGetPythonItem(Name).values()
+
+        # Return the c++ and Python registry values in a tuple
+        # Note that we return None if the item was not present in one of the registries
+        return py_values, cpp_values
+
+    def items(self, Name):
+        # Check if Name registry level is registered and if it is a value
```

Comment:
It should be value?

Reply:
I think the comment is correct. We first check if `Name` is registered and then if it is a value item of a subitem. If the former we throw an error.

In any case, thanks to this observation I found a copy & paste error in the error thrown messages below.

Reply:
I think what he means is that the naming is confusing with `dict`

Reply:
Sorry guys, but I still don't gent your point at all... :confused: :facepalm: